### PR TITLE
Fix TPU kernel OOB memory access 

### PR DESCRIPTION
**Pull request by darisoy**
# Description

This PR fixes a critical TPU hardware crash (Denial of Service / Machine Check Exception) caused by a negative prefetch index in MaxText ragged attention kernels, and establishes full CPU-only CI test coverage using JAX Pallas Interpret mode.

### Proposed Changes & Implementation:

1.  **Negative Prefetch Index Prevention:** Clamped `last_good_block` in `ragged_attention.py` to `0` using `jnp.maximum(0, ...)` to ensure empty sequence paddings safely access valid bounds, preventing out-of-bounds TPU hardware crashes (`CWE-125`).
2.  **Preserved Intermediate Scaling & Documented:** Retained the intermediate `o = o * l` unnormalization inside `ragged_mha` and `ragged_gqa` because the outer MaxText `Attention` layer (`attention_op.py`) explicitly relies on unnormalized outputs to perform multi-chunk prefill/decode cache merging and final normalization. Added clear explanatory inline comments to document this architectural requirement.
3.  **Unified CPU Interpret Mode:** Exposed `interpret` parameter in MHA/GQA wrappers to support CPU-based testing and added full unit test coverage on CPU.

BUG=510376806

---

# Tests

The changes were verified using both live TPU hardware (`tpu7x-8` VM) and CPU-based Pallas Interpret mode.

### 1. TPU-Based Verification
The unit tests in `tests/unit/kernels_test.py` correctly align scales using the division workaround (`ragged_out / ragged_denom`) to verify unnormalized outputs against reference implementations.

Command run inside the virtual environment on TPU VM:
```bash
source maxtext_venv/bin/activate
pytest tests/unit/kernels_test.py -k "not Cpu"
```
**Result:** **`3 passed`** (TPU execution verified).

### 2. CPU-Based Verification & Coverage (Ragged Attention)
To ensure code coverage in CPU-only CI environments, we added a new test suite `RaggedAttentionCpuTest` in `tests/unit/kernels_test.py`. This suite forces JAX Pallas to run in **Interpret Mode** on CPU (bypassing TPU compilation).

Command run locally or in CPU-only CI:
```bash
JAX_PLATFORMS=cpu pytest tests/unit/kernels_test.py -k "Cpu"
```
**Result:** **`3 passed`** (CPU interpret mode verified). This provides full test coverage for `ragged_attention.py` in standard GitHub CI runners.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by darisoy** [comment]
### Note on Code Coverage (SparseCore Kernels)

You will notice that Codecov reports `0%` Patch Coverage for the newly added security and bounds-checking lines in the following two files:
*   `src/maxtext/kernels/gather_reduce_sc.py` (Mosaic SparseCore)
*   `src/maxtext/kernels/gather_reduce_pallas.py` (Pallas SparseCore)

**Why these lines are not covered in the CI:**
1.  **Hardware Requirements:** These SparseCore gather-reduce kernels rely on low-level Mosaic MLIR dialects and SparseCore ASIC hardware, which are only supported on actual TPU instances (TPU v5e/v6e/v7+).
2.  **CPU CI Limitation:** The GitHub CI Code Coverage job runs on standard CPU-only runners. Mosaic MLIR has no CPU interpreter, and Pallas CPU interpret mode does not support or emulate SparseCore hardware features.
3.  **Skipped Tests:** Consequently, the SparseCore unit tests (`tests/gather_reduce_sc_test.py`) have hardware checks that automatically skip them on CPU, meaning these files are never executed during the CI coverage runs.

**Manual Verification:**
All SparseCore clamping, off-by-one, and underflow bounds-checking fixes have been **manually verified and compiled successfully on a live `tpu7x-8` VM**. 

*(Note: The Ragged Attention fixes in `ragged_attention.py` have been fully covered and verified in CI using JAX Pallas's CPU Interpret Mode in `tests/unit/kernels_test.py`).*

File: codecov.yml [hunk 1]
```diff
@@ -38,6 +38,9 @@ ignore:
   - "src/maxtext/experimental"
   - "src/maxtext/inference"
   - "src/maxtext/scratch_code"
+  # Exclude low-level hardware-bound SparseCore TPU kernels that cannot run on CPU-only CI runners
+  - "src/maxtext/kernels/gather_reduce_pallas.py"
+  - "src/maxtext/kernels/gather_reduce_sc.py"
```
Comment (bvandermoon) [inline codecov.yml]: Could tests for these run on the TPU runners?

Comment (darisoy) [reply codecov.yml]: TPU runners use v6e i believe and these kernels appear to be for tpu7x and later per the skip logic in `tests/gather_reduce_sc_test.py`

File: src/maxtext/kernels/gather_reduce_sc.py [hunk 2]
```diff
@@ -128,6 +128,11 @@ def sc_gather_reduce(
     The result of operation, bf16 matrix [M/reduce_group_size, K].
   """
 
+  assert (
+      op.shape[1] % col_chunk_size == 0
+  ), f"op.shape[1] ({op.shape[1]}) must be perfectly divisible by col_chunk_size ({col_chunk_size})"
+  idx = jnp.clip(idx, 0, op.shape[0] - 1)
```
Comment (clee1994) [inline src/maxtext/kernels/gather_reduce_sc.py]: this might have perf implications I assume, so maybe have an options to run this with this clipping? (e.g. there might be cases where the user can guarantee this behavior through other ways?)

**Review by clee1994 — COMMENTED** [review]
LGTM, thanks for fixing this!

File: src/maxtext/kernels/gather_reduce_sc.py (hunk 2 above)
Comment (darisoy) [reply src/maxtext/kernels/gather_reduce_sc.py]: makes sense. hid this behind an `enable_clipping` flag

File: src/maxtext/kernels/attention/ragged_attention.py [hunk 3]
```diff
@@ -325,13 +329,13 @@ def ragged_mha(
...
           mask_value=mask_value,
           cost_estimate=cost_estimate,
+          interpret=interpret,
       ),
       in_axes=(1, 1, 1, None),
       out_axes=2,
   )(query, key, value, lengths)
   m = jnp.expand_dims(m, axis=-1)
   l = jnp.expand_dims(l, axis=-1)
-  o = o * l
```
Comment (NuojCheng) [inline src/maxtext/kernels/attention/ragged_attention.py:334]: why removing this line?

Comment (NuojCheng) [reply src/maxtext/kernels/attention/ragged_attention.py:334]: oh I see, so the output definition of `ragged_mha` gets changed. If there is a correctness test for the kernel and get passed it should be fine.

File: src/maxtext/kernels/gather_reduce_sc.py [hunk 4]
```diff
@@ -237,7 +250,7 @@ def fill_load_offset_tile(offset_tile_local, idx_tile_local, col_pos):
         )
         loop_over_idx = scf.ForOp(
             lower_bound=const_lut(0),
-            upper_bound=const_lut(row_chunk_size + 1),
+            upper_bound=const_lut(row_chunk_size),
```
Comment (NuojCheng) [inline src/maxtext/kernels/gather_reduce_sc.py]: does this line change remove the last iteration?

File: codecov.yml (hunk 1 above)
Comment (NuojCheng) [reply codecov.yml]: these tests should work for v5p as well

**Review by NuojCheng — APPROVED** [review]

**Review by gobbleturk — APPROVED** [review]

File: src/maxtext/kernels/attention/ragged_attention.py (hunk 3 above)
Comment (darisoy) [reply src/maxtext/kernels/attention/ragged_attention.py:334]: oh wait good catch.

I originally removed this because it looks like we were un-normalizing output that was already normalized in https://github.com/AI-Hypercomputer/maxtext/blob/fix-security-vulns/src/maxtext/kernels/attention/ragged_attention.py#L210 which we were normalizing again for the tests to pass. i thought the outer layer was expecting normalized output.

i just did some digging, looks like this was an incorrect assumption. outer layer expects unnormalized output: https://github.com/AI-Hypercomputer/maxtext/blob/fix-security-vulns/src/maxtext/layers/attention_op.py#L2116

Comment (darisoy) [reply src/maxtext/kernels/attention/ragged_attention.py:334]: restored this line and added a comment explaining why it's there

File: src/maxtext/kernels/gather_reduce_sc.py (hunk 4 above)
Comment (darisoy) [reply src/maxtext/kernels/gather_reduce_sc.py]: v5p and tpu7x sparsecore registers have 16 lanes so the index should loop over lanes 0-15. not 0 through 16. hence why the +1 was flagged as a security vulnerability.

Comment (darisoy) [reply src/maxtext/kernels/gather_reduce_sc.py]: turns out the above comment is not accurate and the +1 is indeed needed. reverted this change
